### PR TITLE
Regression tests

### PR DESCRIPTION
**Pull request by chldkdtn**
Additional regression tests are added. Please make sure if all the tests pass in your own platform. Simple go to tests directory and run the following:

./runRegressionTests.sh

You will see which tests fail as the progress is made. A detailed description of how to run the regression tests is given in tests/REGRESSIONTEST.md file.

### CODE REVIEW

**Comment by chldkdtn** [comment]
@steffi7574 and @andersp , we probably want to "squash and merge" this.

**Comment by andersp** [comment]
@chldkdtn did you really mean to change 350 files in this PR?

**Comment by chldkdtn** [comment]
@andersp Those files are output files from the tests that are newly added in this PR. Do you think they are too many? Then, we can remove some of them. If you can suggest which files to remove, I would appreciate it. Or another way is to merge those output files into one file and compare only one file. Just let me know.

**Comment by andersp** [comment]
I took a look at some of the tests. The trusted results always seem to be in a sub-folder called 'base'. Of all the files you save now, it seems only grad.dat and optim_history.dat are needed. I suggest getting rid of the rest.

**Comment by chldkdtn** [comment]
@andersp I can definitely get rid of the rest. For pipulse though, I believe that the gradient computation is not implemented yet. Is it okay to compare only optim_history.dat for pipulse test?

File: tests/cnot/base/optim_history.dat [hunk 1]
```diff
@@ -0,0 +1 @@
+#iter    Objective           ||Pr(grad)||           LS step           1-F_avg         Tikhonov-regul      Penalty-term
```
Comment (steffi7574) [inline tests/cnot/base/optim_history.dat:1]: This file is empty and should be removed. Please also check for other empty files.

**Comment by steffi7574** [comment]
Thanks for the changes so far! A few points:
* Yes the pipulse only runs in 'primal' mode only, no adjoint. For this it would be good to compare the full density matrix at the last time step. 
* Generally, to test the forward propagation, it is enough to compare the **last** time-step only, not the entire time-history (this is for test cases that output of the full density matrix rho, as well as for the expected energy and population). 
* There are a few empty 'optim_history.dat' files which can be removed. 
* Would you mind adding a few words about the regression tests in the user guide?

File: tests/cnot/base/optim_history.dat (hunk 1 above)
Comment (chldkdtn) [reply tests/cnot/base/optim_history.dat:1]: @steffi7574 Thank you for discovery. I am wondering why it is empty although it does compute gradient. I will look into this.

Comment (chldkdtn) [reply tests/cnot/base/optim_history.dat:1]: Not sure what happened, but now it generates non-empty optim_history.dat files, so I have included them.

**Comment by chldkdtn** [comment]
@steffi7574, Now I have included density matrix output files in which only first and last time step density matrices are written only for primal simulation, which do not support adjoint, i.e., pipulse. All the empty optim_history.dat has been replaced with non-empty optim_history.dat. I have added a few sentences in user guide about our regression tests.

**Comment by andersp** [comment]
@steffi7574 @chldkdtn are we ready to merge this PR?

**Review by steffi7574 — APPROVED** [review]

**Pull request merged**